plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p3order/LayerSweepCrossingMinimizer.java

```diff
@@ -179,6 +180,16 @@ private void compareDifferentRandomizedLayouts(final GraphInfoHolder gData) {
         graphsWhoseNodeOrderChanged.clear();
 
         int bestCrossings = Integer.MAX_VALUE;
+        if (gData.lGraph().getProperty(LayeredOptions.PRESERVE_ORDER) != OrderingStrategy.NONE) {
+            // Set the current order of nodes and ports as the currently best order.
+            // If no unnecessary crossings are produced by this solution this order is preserved.
+            bestCrossings = countCurrentNumberOfCrossings(gData);
+            gData.setBestNodeNPortOrder(new SweepCopy(gData.lGraph().toNodeArray()));
+            if (bestCrossings == 0) {
+                return;
```

Comment:
Again, are you really sure that all dummy nodes and edges that may have been introduced at this point have been placed in an acceptable position? If not, I would suggest to run at least one forward sweep. 

If you are sure, please add a corresponding comment at this point in the code.

Reply:
I am quite sure that everything is at the correct place. I checked several models that caused no issue including models with self loops and everything. If they are not, then this should cause additional crossings or should be visible in the graph.

I guess it would be best to not set bestCrossings here and not set the bestNode order and just do one sweep to be sure.

Reply:
Apparently it does make a difference if I sweep one time. Do you have any idea why?
This is the desired layout based on the model order:
![Screenshot from 2020-07-09 16-20-00](https://user-images.githubusercontent.com/6419799/87051553-018c9380-c200-11ea-8aed-d844a4330226.png)
and this is produced if I do not set the best order here:
![Screenshot from 2020-07-09 16-21-28](https://user-images.githubusercontent.com/6419799/87051652-26810680-c200-11ea-9923-9145170e87ae.png)

This seems to be the same issue I had when trying out the idea you mentioned below that should solve this problem in a simpler way.

Reply:
I think the problem lies in the barycenter heuristic which randomizes barycenters of l2p2 and l2p3 since they sources. If I do not randomize in that case it should work, maybe also without sorting ports, but only by sorting nodes.

Reply:
However, I still have to sort ports since there are port constellations in which I purposely deviate from the model order to prevent a randomized solution.
Desired solution:
![Screenshot from 2020-07-09 17-02-39](https://user-images.githubusercontent.com/6419799/87056586-1e2bca00-c206-11ea-9a66-13f30447535c.png)

Solution without ordering edge 3 above edge 2 since edge 1 and edge 3 connect to the same long edge target node:
![Screenshot from 2020-07-09 17-03-14](https://user-images.githubusercontent.com/6419799/87056596-21bf5100-c206-11ea-8e31-0dadf0aeb047.png)

Reply:
I look into the node and port comparator but it seems that there are always edge cases that still need a specific sorting of nodes/dummy nodes and ports.